contrib/scaffolder/template-testing-dry-run.md

```diff
@@ -0,0 +1,168 @@
+};
+
+const writeResultContents = async (
+  root: string,
+  directoryContents: ScaffolderDryRunResponse['directoryContents'],
+) => {
+  const directories = new Set(
+    directoryContents.map(({ path }) => dirname(join(root, path))),
+  );
+  await Promise.all(Array.from(directories).map(d => ensureDir(d)));
+  await Promise.all(
+    directoryContents.map(async ({ path, base64Content, executable }) =>
+      writeFile(join(root, path), Buffer.from(base64Content, 'base64'), {
+        mode: executable ? 0o755 : 0o644,
+      }),
+    ),
+  );
+};
+
+const getToken = async (backendSecret: string) => {
+  const signingKey = base64url.decode(backendSecret);
+  return await new SignJWT({})
+    .setProtectedHeader({ alg: TOKEN_ALG })
+    .setSubject(TOKEN_SUB)
+    .setExpirationTime('10min')
+    .sign(signingKey);
+};
+
+const api = async (
+  bodyObj: Record<string, any>,
+  { baseURL, token }: { baseURL: string; token: string | false },
+) => {
+  const body = gzipSync(JSON.stringify(bodyObj));
+  return fetch(new URL('/api/scaffolder/v2/dry-run', baseURL), {
+    method: 'POST',
+    headers: {
+      Authorization: `Bearer ${token || ''}`,
+      'Content-Type': 'application/json',
+      'Content-Encoding': 'gzip',
+    },
+    body,
+  });
+};
+
+const handle = async (
+  baseURL: string,
+  template_path: string,
+  data: string,
+  target: string,
+  {
+    token,
+    backendSecret,
+  }: { token: string | false; backendSecret: string | false },
+) => {
+  const directoryContents = await loadDirectoryContents(template_path);
+  const values = parse(await readFile(data, 'utf-8'));
+  const secrets = {};
+  const template = parse(
+    await readFile(`${template_path}/template.yaml`, 'utf-8'),
+  );
+  if (backendSecret) {
+    // eslint-disable-next-line no-param-reassign
+    token = await getToken(backendSecret);
+  }
+  const response = await api(
+    {
+      directoryContents,
+      values,
+      secrets,
+      template,
+    },
+    { baseURL, token },
+  );
+  if (!response.ok) {
+    const contentType = response.headers.get('content-type');
+    if (contentType?.startsWith('application/json')) {
+      const responseData = await response.json();
+      if (responseData.error) {
+        throw responseData.error;
+      }
+      if (responseData.errors) {
+        throw responseData.errors;
+      }
+      throw responseData;
+    }
+    throw await response.text();
+  }
+  const { log, directoryContents: resultContents } =
+    (await response.json()) as ScaffolderDryRunResponse;
+  for (const logEntry of log) {
+    console.log(logEntry.body.message);
+  }
+  await writeResultContents(target, resultContents);
+};
+
+const main = async (argv: string[]) => {
+  program
+    .name('backstage-scaffolder')
+    .version(version)
+    .description('Creates a dry-run output of a given template')
+    .option('-t, --token <value>', 'JWT to use for auth')
+    .option('-b, --backend-secret <value>', 'Base64 encoded backend secret')
```

Comment:
Not sure what happened here, could've sworn I left a comment somewhere about this 😅 

I don't want to encourage usage of backend secrets outside of the backend in any way to be honest. This is not a secret that you should be throwing around on dev machines in any way. In an optimal setup, nobody even touched or touches the secret during its lifecycle. Let's skip this option and just go with the token approach for where it's needed. I do want to get to a point where we can have a nicer flow, but we're not there yet.

Reply:
Adjusted. We might back out the back-out internally to do CI, but I see why you wouldn't want to recommend that.

I've also added some `ora` use for user feedback.

Reply:
Yep that's all good! And not gonna try to hide the history or anything, just wanna be careful with our recommendations